crypto/hpke.c

```diff
@@ -0,0 +1,2730 @@
+ *
+ * Note: This also need blanks
+ */
+const char *hpke_kem_strtab[]={
+    NULL,NULL,NULL,NULL,
+    NULL,NULL,NULL,NULL,
+    NULL,NULL,NULL,NULL,
+    NULL,NULL,NULL,NULL,
+    HPKE_KEMSTR_P256,
+    HPKE_KEMSTR_P384,
+    HPKE_KEMSTR_P521,
+    NULL,
+    NULL,NULL,NULL,NULL,
+    NULL,NULL,NULL,NULL,
+    NULL,NULL,NULL,NULL,
+    HPKE_KEMSTR_X25519,
+    HPKE_KEMSTR_X448,
+    NULL};
+#endif
+
+/*!
+ * @brief info about a KDF
+ */
+typedef struct {
+    uint16_t       kdf_id; /**< code point for KDF */
+    const EVP_MD*  (*hash_init_func)(void); /**< the hash alg we're using */
+    size_t         Nh; /**< length of hash/extract output */
+} hpke_kdf_info_t;
+
+/*!
+ * @brief table of KDFs
+ */
+static hpke_kdf_info_t hpke_kdf_tab[]={
+    { 0, NULL, 0 }, /* keep indexing correct */
+    { HPKE_KDF_ID_HKDF_SHA256, EVP_sha256, 32 },
+    { HPKE_KDF_ID_HKDF_SHA384, EVP_sha384, 48 },
+    { HPKE_KDF_ID_HKDF_SHA512, EVP_sha512, 64 }
+};
+
+#if defined(SUPERVERBOSE) || defined(TESTVECTORS)
+/*
+ * @brief table of KDF strings
+ */
+const char *hpke_kdf_strtab[]={
+    NULL,
+    HPKE_KDFSTR_256,
+    HPKE_KDFSTR_384,
+    HPKE_KDFSTR_512};
+#endif
+
+static OSSL_LIB_CTX *hpke_libctx=NULL;
+
+/*!
+ * <pre>
+ * Since I always have to reconstruct this again in my head...
+ * Bash command line hashing starting from ascii hex example:
+ *
+ *    $ echo -e "4f6465206f6e2061204772656369616e2055726e" | \
+ *       xxd -r -p | openssl sha256
+ *    (stdin)= 55c4040629c64c5efec2f7230407d612d16289d7c5d7afcf9340280abd2de1ab
+ *
+ * The above generates the Hash(info) used in Appendix A.2
+ *
+ * If you'd like to regenerate the zero_sha256 value above, feel free
+ *    $ echo -n "" | openssl sha256
+ *    echo -n "" | openssl sha256
+ *    (stdin)= e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+ * Or if you'd like to re-caclulate the sha256 of nothing...
+ *  SHA256_CTX sha256;
+ *  SHA256_Init(&sha256);
+ *  char* buffer = NULL;
+ *  int bytesRead = 0;
+ *  SHA256_Update(&sha256, buffer, bytesRead);
+ *  SHA256_Final(zero_sha256, &sha256);
+ * ...but I've done it for you, so no need:-)
+ * static const unsigned char zero_sha256[SHA256_DIGEST_LENGTH] = {
+ *     0xe3, 0xb0, 0xc4, 0x42, 0x98, 0xfc, 0x1c, 0x14,
+ *     0x9a, 0xfb, 0xf4, 0xc8, 0x99, 0x6f, 0xb9, 0x24,
+ *     0x27, 0xae, 0x41, 0xe4, 0x64, 0x9b, 0x93, 0x4c,
+ *     0xa4, 0x95, 0x99, 0x1b, 0x78, 0x52, 0xb8, 0x55};
+ * </pre>
+ */
+
+/*!
+ * @brief decode ascii hex to a binary buffer
+ *
+ * @param ahlen is the ascii hex string length
+ * @param ah is the ascii hex string
+ * @param blen is a pointer to the returned binary length
+ * @param buf is a pointer to the internally allocated binary buffer
+ * @return 1 for good otherwise bad
+ */
+int hpke_ah_decode(
+        size_t ahlen, const char *ah,
+        size_t *blen, unsigned char **buf)
+{
+    size_t lblen=0;
+    size_t i=0;
+    unsigned char *lbuf=NULL;
+    if (ahlen <=0 || ah==NULL || blen==NULL || buf==NULL) {
```

Comment:
nit: blank line after declaration.
nit : space around operators

Reply:
fixed in 32f0f8a64b10ff8c503b04fe71cbe47a7aabb968

Reply:
fixed in 32f0f8a